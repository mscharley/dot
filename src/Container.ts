import type * as interfaces from './interfaces/index.js';
import type { Binding } from './models/Binding.js';
import type { BindingBuilder } from './BindingBuilder.js';
import { calculatePlan } from './planner/calculatePlan.js';
import { ClassBindingBuilder } from './BindingBuilder.js';
import { executePlan } from './planner/executePlan.js';
import { getConstructorParameterInjections } from './decorators/registry.js';
import { isNever } from './util/isNever.js';
import { noop } from './util/noop.js';
import type { Request } from './models/Request.js';
import type { Token } from './Token.js';
import { tokenForIdentifier } from './util/tokenForIdentifier.js';

export class Container implements interfaces.Container {
	static #currentRequest: Request<unknown> | undefined;

	public static get isProcessingRequest(): boolean {
		return this.#currentRequest != null;
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	readonly #incompleteBindings = new Set<BindingBuilder<any>>();

	#bindings: Array<Binding<unknown>> = [];
	readonly #log: interfaces.Logger[interfaces.LoggerLevel];
	readonly #singletonCache: Record<symbol, unknown> = {};
	public readonly config: Readonly<interfaces.ContainerConfiguration>;

	public constructor(config?: Partial<interfaces.ContainerConfiguration>) {
		this.config = {
			defaultScope: 'transient',
			logLevel: 'debug',
			logger: { info: noop, debug: noop, trace: noop },
			...config,
		};
		this.#log = this.config.logger[this.config.logLevel];
	}

	public static resolve<T>(token: Token<T>): T {
		if (this.#currentRequest == null) {
			throw new Error(
				`Unable to resolve token as no container is currently making a request: ${token.identifier.toString()}`,
			);
		}

		if (!(token.identifier in this.#currentRequest.stack)) {
			throw new Error(`Token hasn't been created yet: ${token.identifier.toString()}`);
		}
		const tokenStack = this.#currentRequest.stack[token.identifier] as T[];
		const [value] = tokenStack.splice(0, 1);
		if (tokenStack.length === 0) {
			delete this.#currentRequest.stack[token.identifier];
		}

		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		return value!;
	}

	#validateBindings = (): void => {
		const values = [...this.#incompleteBindings.values()];
		if (values.length > 0) {
			throw new Error(
				`Some bindings were started but not completed: ${values.map((v) => v.token.identifier.toString()).join(', ')}`,
			);
		}
	};

	public bind = (<T>(id: interfaces.ServiceIdentifier<T>): interfaces.BindingBuilder<T> => {
		const binding = new ClassBindingBuilder(
			id as interfaces.ServiceIdentifier<object>,
			this.config,
			this.addBinding,
		) as unknown as BindingBuilder<T>;
		this.#incompleteBindings.add(binding);

		return binding;
	}) as interfaces.BindFunction;

	public unbind: interfaces.UnbindFunction = (id) => {
		const token = tokenForIdentifier(id);
		const bindings = this.#bindings.flatMap((b) => (b.token.identifier === token.identifier ? [token.identifier] : []));
		if (bindings.length === 0) {
			throw new Error(`Unable to unbind token because it is not bound: ${token.identifier.toString()}`);
		}

		this.#bindings = this.#bindings.filter((b) => !bindings.includes(b.token.identifier));
	};

	public rebind: interfaces.RebindFunction = (<T>(id: interfaces.ServiceIdentifier<T>) => {
		this.unbind(id);
		return this.bind(id);
	}) as interfaces.RebindFunction;

	public load = ((module): void | Promise<void> => {
		return module(this.bind, this.unbind, this.has, this.rebind);
	}) as interfaces.Container['load'];

	public createChild: interfaces.Container['createChild'] = (options) => {
		return new Container({ ...options, parent: this });
	};

	public addBinding = <T>(builder: BindingBuilder<T>, binding: Binding<T>): void => {
		this.#incompleteBindings.delete(builder);
		this.#bindings.push(binding as Binding<unknown>);
	};

	#resolveBinding = <T>(binding: Binding<T>): T | Promise<T> => {
		switch (binding.type) {
			case 'static':
				return binding.value;
			case 'dynamic':
				return binding.generator({ container: this, id: binding.id });
			case 'constructor': {
				const args = getConstructorParameterInjections(binding.ctr)
					// eslint-disable-next-line @typescript-eslint/no-magic-numbers
					.sort((a, b) => (a.index < b.index ? -1 : 1))
					.map(({ token }) => Container.resolve(token));

				return new binding.ctr(...args);
			}
			default:
				return isNever(binding, 'Unknown binding found');
		}
	};

	public get = async <T>(
		id: interfaces.ServiceIdentifier<T>,
		options?: Partial<interfaces.InjectOptions>,
	): Promise<T> => {
		this.#validateBindings();
		const token = tokenForIdentifier(id);

		const plan = calculatePlan<T>(
			this.#bindings,
			this.#resolveBinding,
			{
				type: 'request',
				options: {
					multiple: false,
					optional: false,
					...options,
				},
				token,
			},
			this.config.parent,
		);
		this.#log({ id, options, plan }, 'Processing request');
		const request: Request<T> = {
			stack: {},
			singletonCache: this.#singletonCache,
			token,
		};

		const lastRequest = Container.#currentRequest;
		try {
			Container.#currentRequest = request;
			return await executePlan(plan, request);
		} finally {
			// eslint-disable-next-line require-atomic-updates
			Container.#currentRequest = lastRequest;
		}
	};

	public has: interfaces.IsBoundFunction = (id) => {
		const token = tokenForIdentifier(id);
		return this.#bindings.find((b) => b.token.identifier === token.identifier) != null;
	};
}
