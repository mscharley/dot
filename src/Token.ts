import type * as interfaces from './interfaces/index.js';
import { stringifyIdentifier } from './util/stringifyIdentifier.js';

/**
 * A simple identifier that has a reference to a type which can be injected
 *
 * @remarks
 *
 * Tokens are generated in such a way that each token refers to a unique binding, even if the token is given the same
 * string name. If you use the same name in multiple places then some logs may become a little more confusing but this
 * design prevents name collisions between two potential codebases which may not necessarily know about each other.
 * This allows for libraries which export a ContainerModule and set of tokens as the primary way to interact with the
 * library.
 *
 * @public
 */
export class Token<out T> {
	/**
	 * The Symbol used as the unique identifier for this token
	 *
	 * @internal
	 */
	public readonly identifier: symbol;

	public constructor(name: string) {
		this.identifier = Symbol(name);
	}

	/**
	 * This is a dummy variable, it throws an error on access
	 *
	 * @remarks
	 *
	 * This property exists to facilitate some compatibility across ESM and CommonJS in mixed environments. It is
	 * only here for the extra type information and this property is always undefined.
	 */
	public declare readonly _witness: T;

	/**
	 * Returns a stringified version of this token
	 */
	public toString(): string {
		return stringifyIdentifier(this);
	}

	/**
	 * Returns a simplified version of this token as a JSON blob
	 *
	 * @remarks
	 *
	 * This helps JSON loggers deal with tokens because the only property is a symbol so tokens would show up all tokens
	 * as `{}`.
	 */
	public toJSON(): object {
		return { tokenFor: this.identifier.toString() };
	}
}

/**
 * An identifier that specifies bindings should also have metadata attached to them
 *
 * @remarks
 *
 * See {@link Token} for basic usage guidelines. This is a token which also stipulates some metadata which should be
 * provided along with all bindings for this token.
 *
 * @public
 */
export class MetadataToken<out T, Metadata extends interfaces.MetadataObject> {
	/**
	 * The Symbol used as the unique identifier for this token
	 *
	 * @internal
	 */
	public readonly identifier: symbol;

	public constructor(name: string, public readonly guard: (o: unknown) => o is Metadata) {
		this.identifier = Symbol(name);
	}

	/**
	 * This is a dummy variable, it throws an error on access
	 *
	 * @remarks
	 *
	 * This property exists to facilitate some compatibility across ESM and CommonJS in mixed environments. It is
	 * only here for the extra type information and this property is always undefined.
	 */
	public declare readonly _witness: T;

	/**
	 * Returns a stringified version of this token
	 */
	public toString(): string {
		return stringifyIdentifier(this);
	}

	/**
	 * Returns a simplified version of this token as a JSON blob
	 *
	 * @remarks
	 *
	 * This helps JSON loggers deal with tokens because the only property is a symbol so tokens would show up all tokens
	 * as `{}`.
	 */
	public toJSON(): object {
		return { metadataTokenFor: this.identifier.toString() };
	}
}

/**
 * Helper token type for the common case of simple named bindings
 *
 * @public
 */
export class NamedToken<out T> extends MetadataToken<T, { name: string }> {
	public constructor(name: string) {
		super(name, (o): o is { name: string } => typeof o === 'object' && o != null && 'name' in o && typeof o.name === 'string');
	}
}

/**
 * An easy way to reference any kind of token
 *
 * @public
 */
// eslint-disable-next-line @typescript-eslint/no-type-alias
export type AnyToken<T> = Token<T> | MetadataToken<T, interfaces.MetadataObject>;

/**
 * A helper type to extract the type that a token will inject
 *
 * @public
 */
export type TokenType<T extends { _witness: unknown }> = T extends { _witness: infer U } ? U : never;
