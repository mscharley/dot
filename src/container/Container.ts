import type * as interfaces from '../interfaces/index.js';
import type { Binding, ConstructorBinding } from '../models/Binding.js';
import { getConstructorParameterInjections, getInjections } from '../decorators/registry.js';
import { InvalidOperationError, TokenResolutionError } from '../Error.js';
import type { BindingBuilder } from './BindingBuilder.js';
import { calculatePlan } from '../planner/calculatePlan.js';
import { ClassBindingBuilder } from './BindingBuilder.js';
import { executePlan } from '../planner/executePlan.js';
import type { Injection } from '../models/Injection.js';
import { isNever } from '../util/isNever.js';
import { noop } from '../util/noop.js';
import type { Request } from '../models/Request.js';
import { ResolutionCache } from './ResolutionCache.js';
import { stringifyIdentifier } from '../util/stringifyIdentifier.js';
import type { Token } from '../Token.js';
import { tokenForIdentifier } from '../util/tokenForIdentifier.js';

export class Container implements interfaces.Container {
	static #currentRequest: Request<unknown> | undefined;

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	readonly #incompleteBindings = new Set<BindingBuilder<any>>();

	#bindings: Array<Binding<unknown>> = [];
	readonly #log: interfaces.LoggerFn;
	readonly #warn: interfaces.LoggerFn;
	readonly #singletonCache = new ResolutionCache();
	public readonly config: Readonly<interfaces.ContainerConfiguration>;

	public constructor(config?: Partial<interfaces.ContainerConfiguration>) {
		this.config = {
			autobindClasses: false,
			defaultScope: 'transient',
			logLevel: 'debug',
			logger: { debug: noop, info: noop, trace: noop, warn: noop },
			...config,
		};
		this.#log = this.config.logger[this.config.logLevel].bind(this.config.logger);
		this.#warn = this.config.logger.warn.bind(this.config.logger);
	}

	/**
	 * Attempt to resolve a token using the currently executing request
	 *
	 * @remarks
	 *
	 * There is only a brief window where this is valid, while synchronously constructing a class. Because this relies on
	 * only synchronous processing there is no chance of a race condition in JavaScript despite this relying on some
	 * static shenanigans.
	 */
	public static resolvePropertyInjection<T>(token: Token<T>, resolutionPath: Array<Token<unknown>>): T {
		if (this.#currentRequest == null) {
			throw new InvalidOperationError(
				`Unable to resolve token as no container is currently making a request: ${token.identifier.toString()}`,
			);
		}

		return (this.#currentRequest.container as Container).resolve(token, this.#currentRequest, resolutionPath);
	}

	public resolve<T>(token: Token<T>, request: Request<unknown>, resolutionPath: Array<Token<unknown>>): T {
		if (!(token.identifier in request.stack)) {
			throw new TokenResolutionError(
				'Unable to find a value to inject',
				resolutionPath,
				new InvalidOperationError(`Value for token hasn't been created yet: ${token.identifier.toString()}`),
			);
		}
		const tokenStack = request.stack[token.identifier] as T[];
		const [value] = tokenStack.splice(0, 1);
		if (tokenStack.length === 0) {
			delete request.stack[token.identifier];
		}

		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		return value!;
	}

	#ensureBindingsCompleted = (): void => {
		const values = [...this.#incompleteBindings.values()];
		if (values.length > 0) {
			throw new InvalidOperationError(
				`Some bindings were started but not completed: ${values.map((v) => v.token.identifier.toString()).join(', ')}`,
			);
		}
	};

	public bind = (<T>(id: interfaces.ServiceIdentifier<T>): interfaces.BindingBuilder<T> => {
		const binding = new ClassBindingBuilder(
			id as interfaces.ServiceIdentifier<object>,
			this.config,
			this.#warn,
			this,
		) as unknown as BindingBuilder<T>;
		this.#incompleteBindings.add(binding);

		return binding;
	}) as interfaces.BindFunction;

	public unbind: interfaces.UnbindFunction = (id) => {
		const token = tokenForIdentifier(id);
		this.#singletonCache.flushToken(token);
		const bindings = this.#bindings.flatMap((b) => (b.token.identifier === token.identifier ? [token.identifier] : []));
		if (bindings.length === 0 && !this.has(id)) {
			throw new Error(`Unable to unbind token because it is not bound: ${token.identifier.toString()}`);
		}

		this.#bindings = this.#bindings.filter((b) => !bindings.includes(b.token.identifier));
	};

	public rebind: interfaces.RebindFunction = (<T>(id: interfaces.ServiceIdentifier<T>) => {
		this.unbind(id);
		return this.bind(id);
	}) as interfaces.RebindFunction;

	public load: interfaces.Container['load'] = <M extends interfaces.ContainerModule>(module: M): ReturnType<M> =>
		module(this.bind, this.unbind, this.has, this.rebind) as ReturnType<typeof module>;

	public createChild: interfaces.Container['createChild'] = (options) => {
		return new Container({ ...this.config, ...options, parent: this });
	};

	public addBinding = <T>(builder: BindingBuilder<T>, binding: Binding<T>): void => {
		this.#incompleteBindings.delete(builder);
		this.#bindings.push(binding as Binding<unknown>);
	};

	#getBindings = <T>(id: interfaces.ServiceIdentifier<T>): Array<Binding<T>> => {
		const token = tokenForIdentifier(id);
		const explicitBindings = this.#bindings.filter((b) => b.token.identifier === token.identifier) as Array<Binding<T>>;

		if (explicitBindings.length > 0) {
			return explicitBindings;
		}

		if (this.config.autobindClasses && typeof id === 'function' && !(this.config.parent?.has(id) ?? false)) {
			return [
				{
					type: 'constructor',
					ctr: id,
					id,
					token: tokenForIdentifier(id),
					scope: this.config.defaultScope,
				} satisfies ConstructorBinding<T>,
			];
		}

		return [];
	};

	#resolveBinding =
		(request: Request<unknown>) =>
		<T>(binding: Binding<T>, resolutionPath: Array<Token<unknown>>): T | Promise<T> => {
			const getArgsForParameterInjections = (injections: Array<Injection<unknown>>): unknown[] =>
				injections
					// eslint-disable-next-line @typescript-eslint/no-magic-numbers
					.sort((a, b) => ('index' in a && 'index' in b ? (a.index < b.index ? -1 : 1) : 0))
					.map((i) =>
						i.type === 'unmanagedConstructorParameter'
							? i.value.generator()
							: this.resolve(tokenForIdentifier(i.id), request, resolutionPath),
					);

			switch (binding.type) {
				case 'static':
					return binding.value;
				case 'dynamic': {
					const args = getArgsForParameterInjections(binding.injections);
					return binding.generator(...args);
				}
				case 'constructor': {
					const args = getArgsForParameterInjections(getConstructorParameterInjections(binding.ctr));

					Container.#currentRequest = request;
					try {
						return new binding.ctr(...args);
					} finally {
						Container.#currentRequest = undefined;
					}
				}
				default:
					return isNever(binding, 'Unknown binding found');
			}
		};

	public get = async <T>(
		id: interfaces.ServiceIdentifier<T>,
		options?: Partial<interfaces.InjectOptions>,
	): Promise<T> => {
		this.#ensureBindingsCompleted();

		const request: Request<T> = {
			container: this,
			stack: {},
			singletonCache: this.#singletonCache,
			id,
			token: tokenForIdentifier(id),
		};
		const plan = calculatePlan<T>(
			this.#getBindings,
			this.#resolveBinding(request),
			{
				type: 'request',
				options: {
					multiple: false,
					optional: false,
					...options,
				},
				id,
			},
			[],
			this.config.parent,
		);
		this.#log({ id, options, plan }, 'Processing request');

		return executePlan(plan, request);
	};

	public has: interfaces.IsBoundFunction = (id) => {
		const token = tokenForIdentifier(id);
		const local = this.#bindings.find((b) => b.token.identifier === token.identifier) != null;
		return local || (this.config.parent?.has(id) ?? false) || (typeof id === 'function' && this.config.autobindClasses);
	};

	#validateInjections = (binding: Binding<unknown>, injections: Array<Injection<unknown>>): void => {
		for (const i of injections) {
			if (
				// Optional dependencies are always valid
				i.options.optional ||
				// Unmanaged injections are always valid since they contain the static value to use
				i.type === 'unmanagedConstructorParameter'
			) {
				continue;
			}

			if (!this.has(i.id)) {
				throw new InvalidOperationError(
					`Unbound dependency: ${stringifyIdentifier(binding.id)} => ${stringifyIdentifier(i.id)}`,
				);
			}
		}
	};

	public validate: interfaces.Container['validate'] = (): void => {
		for (const binding of this.#bindings) {
			switch (binding.type) {
				case 'static':
					// These are always valid, the value is in the binding and has no dependencies
					continue;
				case 'dynamic': {
					this.#validateInjections(binding, binding.injections);
					continue;
				}
				case 'constructor': {
					this.#validateInjections(binding, getInjections(binding.ctr));
					continue;
				}
				default:
					return isNever(binding, 'Invalid binding');
			}
		}

		return this.config.parent?.validate();
	};
}
