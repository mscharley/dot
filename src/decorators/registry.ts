import type * as interfaces from '../interfaces/index.js';
import type {
	ConstructorParameterInjection,
	Injection,
	PropertyInjection,
	UnmanagedConstructorParameterInjection,
} from '../models/Injection.js';
import { InvalidOperationError } from '../Error.js';
import { makeGlobalCache } from '../util/makeGlobalCache.js';
import { stringifyIdentifier } from '../util/stringifyIdentifier.js';

const registryCache = Symbol.for('@mscharley/dot:registry-cache');
const registry = makeGlobalCache<interfaces.Constructor<unknown, unknown[]>, Array<Injection<unknown, interfaces.MetadataObject>>>(registryCache);

export const ensureRegistration = <T>(klass: interfaces.Constructor<T>): void => {
	registry.set(klass, registry.get(klass) ?? []);
};

export const registerInjection = <T>(klass: interfaces.Constructor<T>, injection: Injection<unknown, interfaces.MetadataObject>): void => {
	const injections = registry.get(klass) ?? [];
	injections.push(injection);

	registry.set(klass, injections);
};

export const getInjections = <T>(klass: interfaces.Constructor<T>): Array<Injection<unknown, interfaces.MetadataObject>> => {
	const injections = registry.get(klass);
	if (injections == null) {
		throw new InvalidOperationError(`No @injectable() decorator for class: ${stringifyIdentifier(klass)}`);
	}

	return [...injections];
};

export const getPropertyInjections = <T>(klass: interfaces.Constructor<T>): Array<PropertyInjection<unknown, interfaces.MetadataObject>> => {
	const injections = registry.get(klass);
	if (injections == null) {
		throw new InvalidOperationError(`No @injectable() decorator for class: ${stringifyIdentifier(klass)}`);
	}

	return [...injections].filter((i): i is PropertyInjection<unknown, interfaces.MetadataObject> => i.type === 'property');
};

export const getConstructorParameterInjections = <T>(
	klass: interfaces.Constructor<T>,
): Array<ConstructorParameterInjection<unknown, interfaces.MetadataObject> | UnmanagedConstructorParameterInjection<unknown, interfaces.MetadataObject>> => {
	const injections = registry.get(klass);
	if (injections == null) {
		throw new InvalidOperationError(`No @injectable() decorator for class: ${stringifyIdentifier(klass)}`);
	}

	return [...injections].filter(
		(i): i is ConstructorParameterInjection<unknown, interfaces.MetadataObject> | UnmanagedConstructorParameterInjection<unknown, interfaces.MetadataObject> =>
			i.type === 'constructorParameter' || i.type === 'unmanagedConstructorParameter',
	);
};

export const getRegistry = (): ReadonlyMap<interfaces.Constructor<unknown, unknown[]>, Array<Injection<unknown, interfaces.MetadataObject>>> =>
	registry;
