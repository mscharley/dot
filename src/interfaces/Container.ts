/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

import type { ContainerModule, ContainerModuleMeta } from './ContainerModule.js';
import type { IsBoundFunction, UnbindFunction } from './Functions.js';
import type { MetadataForIdentifier, ServiceIdentifier } from './ServiceIdentifier.js';
import type { ContainerConfiguration } from './ContainerConfiguration.js';
import type { ContainerFactory } from './ContainerFactory.js';
import type { InjectedType } from './InjectionIdentifier.js';
import type { InjectOptions } from './InjectOptions.js';

/**
 * An IOC container
 *
 * @remarks
 *
 * You can either manage bindings directly on a container or use {@link @mscharley/dot#interfaces.ContainerModule | ContainerModules }
 * to manage bindings in a reusable and modular way.
 *
 * @public
 */
export interface Container {
	/**
	 * The configuration used by this container
	 */
	readonly config: ContainerConfiguration;

	/** {@inheritdoc interfaces.UnbindFunction} */
	readonly unbind: UnbindFunction;
	/** {@inheritdoc interfaces.IsBoundFunction} */
	readonly has: IsBoundFunction;

	/**
	 * Load a preconfigured module into this container
	 */
	readonly load: <M extends ContainerModule>(module: M) => ReturnType<M>;
	/**
	 * Create a child container
	 */
	readonly createChild: ContainerFactory;
	/**
	 * Fetch a list of all the container modules necessary to satisfy a particular set of requests
	 *
	 * @remarks
	 *
	 * This is mostly useful for build systems to generate partial containers to satisfy particular entrypoints.
	 *
	 * @beta
	 */
	readonly getRequiredContainerModules: (injections: ReadonlyArray<ServiceIdentifier<unknown>>) => ContainerModuleMeta[];

	/**
	 * Make a request from this container
	 */
	readonly get: {
		/* eslint-disable @typescript-eslint/unified-signatures */
		<Id extends ServiceIdentifier<unknown>>(
			id: Id,
			options: { multiple: true } & Partial<InjectOptions<MetadataForIdentifier<Id>>>,
		): Promise<InjectedType<[Id, typeof options]>>;
		<Id extends ServiceIdentifier<unknown>>(
			id: Id,
			options: { optional: true } & Partial<InjectOptions<MetadataForIdentifier<Id>>>,
		): Promise<InjectedType<[Id, typeof options]>>;
		<Id extends ServiceIdentifier<unknown>>(
			id: Id,
			options?: Partial<InjectOptions<MetadataForIdentifier<Id>>>,
		): Promise<InjectedType<[Id, typeof options]>>;
		/* eslint-enable @typescript-eslint/unified-signatures */
	};

	/**
	 * Check that all bindings have no missing dependencies
	 *
	 * @example
	 *
	 * ```typescript
	 * const container = createContainer();
	 * const Name = new Token<string>();
	 *
	 * @injectable(Name)
	 * class Greeter {
	 *   public constructor(private name: string) { }
	 *
	 *   public greet() {
	 *     console.log(`Hello, ${this.name}`);
	 *   }
	 * }
	 *
	 * container.bind(Greeter).toSelf();
	 * container.validate(); // Throws an exception
	 *
	 * container.bind(Name).toConstantValue("John Doe");
	 * container.validate(); // Works
	 * ```
	 */
	readonly validate: (validateAutobindings?: boolean) => void;
}
