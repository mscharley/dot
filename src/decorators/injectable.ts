import { getPropertyInjections, registerInjection } from './registry.js';
import { Container } from '../Container.js';
import type { Injection } from '../models/Injection.js';

/** @public */
export interface InjectableDecorator<T> {
	(target: new () => T, context: ClassDecoratorContext<new () => T>): undefined;
	(target: new () => T, context?: undefined): undefined | (new () => T);
}

const _injections: Injection[] = [];
export const addInjection = (injection: Injection): void => {
	_injections.push(injection);
};

/**
 * @public
 */
export const injectable = <T>(): InjectableDecorator<T> =>
	((target, context?) => {
		/* c8 ignore start */
		if (context == null) {
			// experimental
			// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
			const klass = (() =>
				class extends target {
					public constructor() {
						// eslint-disable-next-line @typescript-eslint/no-unsafe-call
						super();
						getPropertyInjections(klass).forEach(({ name, token }) => {
							// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
							(this as any)[name] = Container.resolve(token);
						});
					}
				})();
			_injections.splice(0).forEach((injection) => {
				registerInjection(klass, injection);
			});

			return klass;
			/* c8 ignore end */
		} else {
			// tc39 - no op
			_injections.splice(0).forEach((injection) => {
				registerInjection(target, injection);
			});

			return undefined;
		}
	}) as InjectableDecorator<T>;
