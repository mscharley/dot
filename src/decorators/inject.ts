/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

import type * as interfaces from '../interfaces/index.js';
import { addInjection } from './injectable.js';
import { Container } from '../container/Container.js';
import { tokenForIdentifier } from '../util/tokenForIdentifier.js';

/**
 * Typesafe definition of a class field decorator
 *
 * @public
 */
export interface ClassFieldDecorator<T extends object, Property> {
	// TC39 definition
	(target: undefined, context: ClassFieldDecoratorContext<T, Property>): (originalValue: Property | undefined) => Property;
	// experimental decorators definition
	(target: T, propertyName: string | symbol): undefined;
}

/**
 * Type for the `@inject` decorator itself
 *
 * @remarks
 *
 * See {@link inject | @inject}
 *
 * @public
 */
export interface InjectDecoratorFactory {
	<T>(id: interfaces.ServiceIdentifier<T>, options: Partial<interfaces.InjectOptions<interfaces.MetadataForIdentifier<typeof id>>> & { multiple: true }): ClassFieldDecorator<object, T[]>;
	<T>(id: interfaces.ServiceIdentifier<T>, options: Partial<interfaces.InjectOptions<interfaces.MetadataForIdentifier<typeof id>>> & { optional: true }): ClassFieldDecorator<object, T | undefined>;
	<T>(id: interfaces.ServiceIdentifier<T>, options?: Partial<interfaces.InjectOptions<interfaces.MetadataForIdentifier<typeof id>>>): ClassFieldDecorator<object, T>;
}

/**
 * Decorator for property injections
 *
 * @example
 *
 * ```typescript
 * const Name = new Token<string>("name");
 *
 * @injectable()
 * class Greeter {
 *   @inject(Name)
 *   public readonly name: string;
 *
 *   public hello() {
 *     console.log(`Hello, ${name}.`);
 *   }
 * }
 * ```
 *
 * @public
 */
export const inject: InjectDecoratorFactory = <T>(
	id: interfaces.ServiceIdentifier<T>,
	options?: Partial<interfaces.InjectOptions<interfaces.MetadataForIdentifier<typeof id>>>,
): ClassFieldDecorator<object, T> => {
	const opts: interfaces.InjectOptions<interfaces.MetadataForIdentifier<typeof id>> = {
		multiple: false,
		optional: false,
		metadata: {},
		...options,
	};
	const token = tokenForIdentifier(id);

	return ((target, context) => {
		/* c8 ignore start */
		if (target != null) {
			// experimental
			addInjection({
				type: 'property',
				name: context as Exclude<typeof context, ClassFieldDecoratorContext<unknown, T>>,
				id: id,
				options: opts,
			});

			return undefined;
			/* c8 ignore end */
		} else {
			// tc39
			addInjection({
				type: 'property',
				name: (context as ClassFieldDecoratorContext<unknown, T>).name,
				id: id,
				options: opts,
			});

			return (_originalValue: T | undefined): T => {
				const value = Container.resolvePropertyInjection<T>(token, [token]);

				return value;
			};
		}
	}) as ClassFieldDecorator<object, T>;
};
