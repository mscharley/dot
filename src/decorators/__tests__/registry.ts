import { beforeEach, describe, expect, it } from '@jest/globals';
import { getInjections, getPropertyInjections, registerInjection } from '../registry.js';
import { Container } from '../../Container.js';
import { inject } from '../inject.js';
import { injectable } from '../injectable.js';
import { Token } from '../../Token.js';

const strToken = new Token<string>('registry.string');

@injectable()
class TestClass {
	@inject(strToken)
	public str!: string;
}
registerInjection(TestClass, {
	type: 'request',
	options: { optional: false },
	token: new Token<TestClass>('registry.class'),
});

describe('registry', () => {
	let container: Container;

	beforeEach(() => {
		container = new Container();
		container.bind(strToken).toConstantValue('hello world');
	});

	describe('getInjections', () => {
		it('can get all injections registered', () => {
			expect(getInjections(TestClass)).toMatchSnapshot();
		});

		it('gets an empty array for a random class', () => {
			expect(getInjections(class {})).toStrictEqual([]);
		});
	});

	describe('getPropertyInjections', () => {
		it('can get all property injections', () => {
			expect(getPropertyInjections(TestClass)).toMatchSnapshot();
		});

		it('gets an empty array for a random class', () => {
			expect(getPropertyInjections(class {})).toStrictEqual([]);
		});
	});
});
