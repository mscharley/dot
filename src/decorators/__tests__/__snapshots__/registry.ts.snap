// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registry getInjections can get all injections registered 1`] = `
[
  {
    "name": "str",
    "options": {
      "optional": false,
    },
    "token": Token {
      "identifier": Symbol(registry.string),
    },
    "type": "property",
  },
  {
    "options": {
      "optional": false,
    },
    "token": Token {
      "identifier": Symbol(registry.class),
    },
    "type": "request",
  },
]
`;

exports[`registry getPropertyInjections can get all property injections 1`] = `
[
  {
    "name": "str",
    "options": {
      "optional": false,
    },
    "token": Token {
      "identifier": Symbol(registry.string),
    },
    "type": "property",
  },
]
`;
